### PR TITLE
added score-addon to compute placement

### PR DESCRIPTION
**Pull request by mzeevi**
With this PR, a new add-on has been introduced, which is responsible for creating a score for each Managed Cluster, so that the suitable cluster for RCS workload can be picked based on CPU requestor Memory request instead of on allocatable CPU and memory. In addition, the quickstart and README.md have been fixed

### CODE REVIEW

File: addons/score/spoke/calculate.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+		memScore = int64(minScore)
+	} else {
+		memScore = int64(200*availableMem/maxMemoryCountNum - 100)
+	}
+
+	klog.Infof("cpuScore = %v, memScore = %v", cpuScore, memScore)
+	return cpuScore, memScore, nil
+}
+
+func (s *Score) calculateClusterAllocateable(resourceName clusterv1.ResourceName) (float64, error) {
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:93]: missing a lot of logs

File: addons/score/spoke/calculate.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+	nodes, err := s.nodeLister.List(labels.Everything())
+	if err != nil {
+		return 0, err
+	}
+
+	allocatableList := make(map[clusterv1.ResourceName]resource.Quantity)
+	for _, node := range nodes {
+		if node.Spec.Unschedulable {
+			continue
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:102]: Its bad practice to use  continue ... change the if structure to `if  !node.spec.Unschedulable`

File: addons/score/spoke/calculate.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+	}
+}
+
+func (s *Score) getMaxMinValues() (float64, float64, float64, float64, error) {
+	maxCPUCount := os.Getenv("MAX_CPU_COUNT")
+	if len(maxCPUCount) == 0 {
+		maxCPUCount = defaultMaxCPUCount
+	}
+
+	maxCPUCountNum, err := strconv.ParseFloat(maxCPUCount, 64)
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:184]: stop using magic numbers  youre not harry potter

File: addons/score/spoke/calculate.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+		quantity, found := (*requests)[resource]
+		if !found {
+			return 0
+		}
+		return quantity.AsApproximateFloat64()
+	}
+}
+
+func (s *Score) getMaxMinValues() (float64, float64, float64, float64, error) {
+	maxCPUCount := os.Getenv("MAX_CPU_COUNT")
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:179]: make a loop for the vars dont just duplicate

**Review by fran2810 — COMMENTED** [review]
Bad code improve it

File: README.md [hunk 5]
```diff
@@ -205,7 +230,43 @@ NAME                                AVAILABLE   DEGRADED   PROGRESSING
...
+  - name: MAX_CPU_COUNT
+    value: "<MAX_CPU_COUNT>"
+  - name: MIN_CPU_COUNT
+    value: "<MIN_CPU_COUNT>>"
+  - name: MAX_MEM_COUNT
+    value: "<MAX_MEM_COUNT>"
+  - name: MIN_MEM_COUNT
+    value: "<MIN_MEM_COUNT>"
+```
+
```
[the remark this answers is not in the document]
Comment (mzeevi) [reply README.md:261]: The entire `README.md` is made up of <PLACEHOLDERS> so we can keep it like this. The `make local-quickstart` deploys this as well with example values.

File: addons/score/spoke/calculate.go [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+// getRequestForResource retrieves the resource request value for the specified resource from the given list of requests.
+// It returns 0 if the requests pointer is nil or if the request for the specified resource is not found.
+func (s *ResourceScore) getRequestForResource(resource corev1.ResourceName, requests *corev1.ResourceList) float64 {
+	if requests == nil {
+		return 0
+	}
+
+	switch resource {
+	case corev1.ResourceCPU:
+		cpuQuantity, found := (*requests)[corev1.ResourceCPU]
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:196]: i dont know what this does

File: addons/score/spoke/calculate.go [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+
+	maxCPUCountNum := maxMinValues[0]
+	minCPUCountNum := maxMinValues[1]
+	maxMemoryCountNum := maxMinValues[2]
+	minMemoryCountNum := maxMinValues[3]
+
+	availableCPU := cpuAllocation - cpuUsage
+	availableMem := memoryAllocation - memoryUsage
+
+	cpuScore := s.calculateNormalizedScore(availableCPU, maxCPUCountNum, minCPUCountNum)
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:94]: cast it to int already here

File: addons/score/spoke/calculate.go [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+	return quantity.AsApproximateFloat64(), nil
+}
+
+// calculatePodResourceRequest returns the total requested quantity of a certain resource by summing up
+// the request of all containers in the cluster by looping over all pods.
+func (s *ResourceScore) calculatePodResourceRequest(resourceName corev1.ResourceName) (float64, error) {
+	var podRequest float64
+	var podCount int
+
+	list, err := s.podListener.List(labels.Everything())
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:156]: list of what .....

File: addons/score/spoke/calculate.go [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+		if !found {
+			return 0
+		}
+		return quantity.AsApproximateFloat64()
+	}
+}
+
+// getMaxMinValues retrieves the maximum and minimum values for CPU and memory counts from environment variables.
+// If the environment variables are not set, it uses default values.
+func (s *ResourceScore) getMaxMinValues() ([]float64, error) {
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:218]: ill return already a map and slice of float

File: addons/score/spoke/calculate.go [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+	envVars := map[string]string{
+		"MAX_CPU_COUNT": defaultMaxCPUCount,
+		"MIN_CPU_COUNT": defaultMinCPUCount,
+		"MAX_MEM_BYTES": defaultMaxMemoryBytes,
+		"MIN_MEM_BYTES": defaultMinMemoryBytes,
+	}
+
+	values := make([]float64, len(envVars))
+
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:227]: this function is a bit problematic you need to pass it the given env vars already parsed

File: addons/score/spoke/calculate.go [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+		valueFloat, err := strconv.ParseFloat(valueToParse, bitSizeFloat)
+		if err != nil {
+			return nil, err
+		}
+
+		values[i] = valueFloat
+		i++
+	}
+
+	return values, nil
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:245]: change the logic of this function to what we talked about

File: addons/score/spoke/calculate.go [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+// If the environment variables are not set, it uses default values.
+func (s *ResourceScore) getMaxMinValues() ([]float64, error) {
+	envVars := map[string]string{
+		"MAX_CPU_COUNT": defaultMaxCPUCount,
+		"MIN_CPU_COUNT": defaultMinCPUCount,
+		"MAX_MEM_BYTES": defaultMaxMemoryBytes,
+		"MIN_MEM_BYTES": defaultMinMemoryBytes,
+	}
+
+	values := make([]float64, len(envVars))
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:226]: bad name values of what ...

**Review by fran2810 — COMMENTED** [review]
make sure you think about `SOLID` when you write you code improve the naming. and function should be easy to unit tests some of your function soes to many things

File: addons/score/spoke/calculate.go (hunk 6 above)
Comment (mzeevi) [reply addons/score/spoke/calculate.go:196]: It references a pointer, I'll change it to make it clearer

File: test/e2e_tests/placement.go [hunk 13]
```diff
@@ -71,4 +72,28 @@ var _ = Describe("Validate the placement controller", func() {
...
 	})
+
+	It("Should update the score of the AddOnPlacementScore", func() {
+		baseCapp := mock.CreateBaseCapp()
+		baseCapp.Spec.Site = testconsts.Cluster2
+		desiredCapp := utilst.CreateCapp(k8sClient, baseCapp)
+
+		By("Checking the score of the AddOnPlacementScore in chosen site")
+		addonPlacementScore := &clusterv1alpha1.AddOnPlacementScore{}
+		Eventually(func() int32 {
```
Comment (fran2810) [inline test/e2e_tests/placement.go:83]: int64

File: test/e2e_tests/placement.go [hunk 14]
```diff
@@ -71,4 +72,28 @@ var _ = Describe("Validate the placement controller", func() {
...
+			addonPlacementScore = utilst.GetAddOnPlacementScore(k8sClient, testconsts.AddOnPlacementScoresName, testconsts.Cluster2)
+			return addonPlacementScore.Status.Scores[0].Value
+		}, testconsts.Timeout, testconsts.Interval).ShouldNot(BeNil(), "Should have a score")
+		score := addonPlacementScore.Status.Scores[0].Value
+
+		By("Checking unique creation of Capp")
```
Comment (fran2810) [inline test/e2e_tests/placement.go:89]: dont know why this check is needed

File: addons/score/spoke/utils.go [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+import (
+	"os"
+	"strconv"
+)
+
+const bitSizeFloat = 64
+
+// getEnvVars returns a map of key values where the keys are passed as a parameter
+// and the values are from environment variables
+func getEnvVars(keys []string) map[string]string {
```
Comment (fran2810) [inline addons/score/spoke/utils.go:12]: name is to generic its extract envVars as map for specific keys `extractEnvAsMap` would be more suitable for this function

File: addons/score/spoke/utils.go [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"os"
+	"strconv"
+)
+
+const bitSizeFloat = 64
+
+// getEnvVars returns a map of key values where the keys are passed as a parameter
+// and the values are from environment variables
```
Comment (fran2810) [inline addons/score/spoke/utils.go:11]: . at the end of sentence.

File: addons/score/spoke/calculate.go [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	return allocation, usage, nil
+}
+
+// normalizeScore returns the normalized score of the CPU and Memory based on the passed values.
+// A valid score must be in the range minScore to maxScore, and there is a need to normalize the
+// scores before updating it into AddOnPlacementScore.
+func (s *ResourceScore) normalizeScore(cpuAllocation, cpuUsage, memoryAllocation, memoryUsage float64) (int64, int64, error) {
+	s.logger.Info("Computing normalized score from resource allocation and usage", "cpuAllocation", cpuAllocation, "cpuUsage", cpuUsage, "memoryAllocation", memoryAllocation, "memoryUsage", memoryUsage)
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:80]: do it in 2 lines it really hard to read coode like this

File: addons/score/spoke/calculate.go [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+		if !found {
+			return 0
+		}
+		return quantity.AsApproximateFloat64()
+	}
+}
+
+// getMaxMinValues retrieves the maximum and minimum values for CPU and memory counts from environment variables.
+// If the environment variables are not set, it uses default values.
+func (s *ResourceScore) getMaxMinValues(maxCPUCount, minCPUCount, maxMemoryBytes, minMemoryBytes string) (map[string]float64, error) {
```
Comment (fran2810) [inline addons/score/spoke/calculate.go:214]: very clean nice

**Review by fran2810 — COMMENTED** [review]
few minor issues good job.

File: test/e2e_tests/placement.go (hunk 14 above)
Comment (mzeevi) [reply test/e2e_tests/placement.go:89]: It serves to make sure the Capp is created

**Review by fran2810 — APPROVED** [review]
/lgtm

**Pull request merged**